lib/prelude.dx

```diff
@@ -2115,6 +2144,41 @@ def concat {n a} (lists:n=>(List a)) : List a =
         eltIdx := eltIdxVal + 1
         xs.(eltIdxVal@_)
 
+def cat_maybes {a n} (xs:n=>Maybe a) : List a =
+  (num_res, res_inds) = yield_state (0:Int, for i:n. Nothing) \ref.
+    for i. case xs.i of
+      Just _ ->
+        ix = get $ fst_ref ref
+        (snd_ref ref) ! (unsafe_from_ordinal _ ix) := Just i
+        fst_ref ref := ix + 1
+      Nothing -> ()
+  to_list $ for i:(Fin num_res).
+    case res_inds.(unsafe_from_ordinal _ $ ordinal i) of
+      Just j -> case xs.j of
+        Just x -> x
+        Nothing -> todo -- Impossible
+      Nothing -> todo -- Impossible
+
+def prev_ix {n} [Ix n] (i:n) : Maybe n =
+  ord = ordinal i
+  if ord == 0
+    then Nothing
+    else Just $ (ord - 1) @ n
```

Comment:
Use `unsafe_from_ordinal` here? You've already checked all the preconditions, haven't you?